src/main/java/uk/nhs/digital/nhsconnect/lab/results/inbound/NhsAckProducerService.java

```diff
@@ -0,0 +1,157 @@
+import uk.nhs.digital.nhsconnect.lab.results.model.edifact.message.MessageParsingException;
+import uk.nhs.digital.nhsconnect.lab.results.sequence.SequenceService;
+import uk.nhs.digital.nhsconnect.lab.results.utils.TimestampService;
+import uk.nhs.digital.nhsconnect.lab.results.inbound.MessageProcessingResult.Success;
+import uk.nhs.digital.nhsconnect.lab.results.inbound.MessageProcessingResult.Error;
+
+import java.time.Instant;
+import java.time.ZoneId;
+import java.time.format.DateTimeFormatter;
+import java.util.ArrayList;
+import java.util.List;
+
+import static uk.nhs.digital.nhsconnect.lab.results.utils.TemplateUtils.loadTemplate;
+import static uk.nhs.digital.nhsconnect.lab.results.utils.TemplateUtils.fillTemplate;
+
+@Component
+@RequiredArgsConstructor(onConstructor_ = @Autowired)
+public class NhsAckProducerService {
+
+    private static final String MESSAGE_ACCEPTED_STATUS_CODE = "MA";
+    private static final String MESSAGE_REJECTED_STATUS_CODE = "MR";
+
+    private static final int MESSAGE_COMMON_SEGMENT_COUNT = 5;
+
+    private final TimestampService timestampService;
+
+    private final SequenceService sequenceService;
+
+    private Mustache nhsAckTemplate = loadTemplate("nhsAck.mustache");
+
+    public String createNhsAckEdifact(WorkflowId workflowId, Interchange interchange,
+                                      List<MessageProcessingResult> messageProcessingResults) {
+
+        InterchangeHeader interchangeHeader = interchange.getInterchangeHeader();
+
+        var nhsAckContent = NhsAckContent.builder()
+            .interchangeSender(interchangeHeader.getSender())
+            .interchangeRecipient(interchangeHeader.getRecipient())
+            .interchangeControlReference(String.valueOf(interchangeHeader.getSequenceNumber()))
+            .nhsAckControlReference(String.valueOf(sequenceService.generateInterchangeSequence(
+                interchangeHeader.getRecipient(), interchangeHeader.getSender())))
+            .workflowId(workflowId.getWorkflowId())
+            .messageStatus(new ArrayList<>()).build();
+
+        int acceptedMessages = 0;
+        int rejectedMessages = 0;
+        int totalMessages = messageProcessingResults.size();
+
+        for (MessageProcessingResult messageProcessingResult : messageProcessingResults) {
+            NhsAckMessageContent nhsAckMessageContent;
+            if (messageProcessingResult instanceof MessageProcessingResult.Success) {
+                Success currentMessageResult = (Success) messageProcessingResult;
+                MessageHeader messageHeader = currentMessageResult.getMessage().getMessageHeader();
+                nhsAckMessageContent = NhsAckMessageContent.builder()
+                        .messageNumber(messageHeader.getSequenceNumber())
+                        .messageStatus(MESSAGE_ACCEPTED_STATUS_CODE)
+                        .build();
+                acceptedMessages++;
+            } else {
```

Comment:
At the moment there are 2 subclasses. After some time someone might want to add another one and it will not work as expected in this `else`. In such cases use explicit `else if` to check other type and add `else { throw ... }` at the end

Reply:
I've made this change.

I added a new class, MessageResultException, that returns the MessageProcessingResult that caused the exception to make it easier to identify the problem.